[WIP] Update VM based templates with private IP and update Tiltfile

nawazkh opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**:
- This PR updates the templates (based off of VMs) to use private IP of the API Server of the control plane of the workload cluster for inter-node communication.
- This PR also updates the Tiltfile with logic to peer AKS VNets with workload cluster's VNets.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/5261

**Special notes for your reviewer**:
<!-- Refer to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/developers/releasing.md#release-support for more information about which changes are eligible for backport -->

- [ ] cherry-pick candidate

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Update VM based templates to use Private IP solution for workload cluster's inter-node communication. Also update Tiltfile to update AKS Vnet with workload cluster's VNet
```

nawazkh commented:
/test pull-cluster-api-provider-azure-e2e

nawazkh commented:
/hold for tests

mboersma reviewed (commented):
This seems like a reasonable approach. Should we have somewhere to document which CIDR ranges are "reserved" by which flavors? If we add a new flavor, we'd want to make sure it doesn't overlap, and it would be better not to have that rely just on maintainers' memories.

nawazkh commented:
Lemme break this PR into multiple smaller PRs updating each of the templates. I will ship this PR with changes to default template and addressing the comments.

nawazkh commented on Tiltfile:
```diff
@@ -317,9 +319,14 @@ def flavors():
     local_resource(
         name = "delete-all-workload-clusters",
-        cmd = kubectl_cmd + " delete clusters --all --wait=false",
+        cmd = ["sh", "-ec", delete_all_workload_clusters],
```
I think we do need `sh` here. We are updating `delete_all_workload_clusters` to have a bunch of delete vnet peering commands added in this function https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/5248/files#diff-c2ee8653e1d6b85f0aadf87cd438a9250806c052877248442be4d434cbc52425R522-R532

nawazkh commented on Tiltfile:
```diff
@@ -382,17 +389,30 @@ def deploy_worker_templates(template, substitutions):
         flavor_cmd += "; until " + kubectl_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig get configmap kubeadm-config --namespace=kube-system > /dev/null 2>&1; do sleep 5; done"
         flavor_cmd += "; " + kubectl_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig create namespace calico-system --dry-run=client -o yaml | " + kubectl_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig apply -f -; " + kubectl_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig get configmap kubeadm-config --namespace=kube-system -o yaml | sed 's/namespace: kube-system/namespace: calico-system/ ...
 
+    # TODO: no one is clearing MGMT_CLUSTER_NAME when using KIND, so this is always going to be true. Improve this logic.
```
Yes, good call out, will update the epic with it.

nawazkh replied:
Created https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/5289

nawazkh commented:
> Should we have somewhere to document which CIDR ranges are "reserved" by which flavors? If we add a new flavor, we'd want to make sure it doesn't overlap, and it would be better not to have that rely just on maintainers' memories.

Where could be a good place to put this?
Does a readme in /templates/flavors be a good place to put the CIDR matrix ?

nawazkh commented:
<img width="1259" alt="Screenshot 2024-11-18 at 3 33 46 PM" src="https://github.com/user-attachments/assets/a5dc2f5d-b31a-4f09-afb8-832c62843394">
Something is up with GitHub, it is not letting me see my changes where I dropped my commits to just 5 instead of 23.

nawazkh commented:
/close

nawazkh commented:
> > Should we have somewhere to document which CIDR ranges are "reserved" by which flavors? If we add a new flavor, we'd want to make sure it doesn't overlap, and it would be better not to have that rely just on maintainers' memories.
> 
> Where could be a good place to put this? Does a readme in /templates/flavors be a good place to put the CIDR matrix ?

Updated https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/5288/files#diff-e7ce6ca4dcc711afa1a2e1afdb50d9c9902749a29f4af50cde1a12f2e51f4858 with the CIDR matrix
